### PR TITLE
Persist state history for resources

### PR DESCRIPTION
**Pull request by robfletcher**
Persists the resource state every time it is checked. I'm also persisting the instance that did the check in the SQL implementation. Can put some metrics around that later perhaps to help track utilization.

I'd like to change the resource state stuff in a future PR so it's persisting a log rather than just the current state.

### CODE REVIEW

File: keel-api/keel-api.gradle [hunk 1]
```diff
@@ -7,6 +7,7 @@ dependencies {
   api project(":keel-core")
   api project(":keel-plugin")
 
+  implementation project(":keel-eureka")
```
Comment (robfletcher) [inline keel-api/keel-api.gradle]: New module

**Comment by robfletcher** [comment]
Actually I think persisting the instance id is bullshit. Let me rework this a bit.

**Comment by cfieber** [comment]
seems like it would be useful info for trying to debug things later if nothing else..

**Comment by robfletcher** [comment]
That's why I started doing it but then I thought what we should really do is emit a spectator metric.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ResourceController.kt [hunk 2]
```diff
@@ -86,10 +90,21 @@ class ResourceController(
...
-    log.info("Deleting: $name")
+    log.debug("Deleting: $name")
     return resourcePersister.handle(ResourceDeleted(name))
   }
 
+  @GetMapping(
+    path = ["/{name}/history"],
+    produces = [APPLICATION_JSON_VALUE, APPLICATION_YAML_VALUE]
+  )
+  fun history(@PathVariable("name") name: ResourceName): List<ResourceStateHistoryEntry> {
```
Comment (robfletcher) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ResourceController.kt]: New endpoint to get state history

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ResourceStateHistoryEntry.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.netflix.spinnaker.keel.persistence
+
+import java.time.Instant
+
+data class ResourceStateHistoryEntry(
+  val state: ResourceState,
+  val timestamp: Instant
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ResourceStateHistoryEntry.kt]: I figure we can add more to this type. The details of the detected diff, for example.

**Comment by robfletcher** [comment]
I'm leaving the Eureka stuff in because the next thing I want to do is Eureka activation so I'll just end up re-implementing it.

**Pull request merged**